extensions/resume.md

```diff
@@ -0,0 +1,162 @@
+---
+title: IRCv3 `resume` extension
+layout: spec
+work-in-progress: true
+copyrights:
+  -
+    name: Daniel Oakley
+    period: 2017
+    email: daniel@danieloaks.net
+---
+## Introduction
+
+Occasionally, clients disconnect from IRC. What happens these days is that the client connects with a different nick, joins all their old channels again, waits for the old connection to time out (or manually kills it using services), and then changes back to their original nickname.
+
+This feature intends to vastly simplify this form of reconnection, and reduce the amount of nick-switching, `JOIN` and `QUIT` notices, and general disruption that other clients see when this happens. At the same time, one vital piece of information to convey on reconnection is that chat history may have been lost. This feature conveys this and also makes such a notice more specific than it currently is (by allowing reconnecting clients to specify exactly how much history they may have lost).
+
+
+## Architecture
+
+This feature is enabled using the `draft/resume` capability, and uses various new messages and numerics to convey state about reconnecting clients.
+
+### Capabilities
+
+The `draft/resume` capability is advertised by servers that support resuming connections. This capability is requested by clients that reconnect in this way with the `RESUME` command and have the ability to correctly parse the `RESUMED` message as described in this specification.
+
+### Messages
+
+There are two new messages defined by this specification:
+
+- `RESUME`: Indicates that the client wishes to resume a previous connection.
+- `RESUMED`: Indicates that a client has disconnected+reconnected to the server.
+
+Below, we go through both of these.
+
+#### `RESUME` Command
+
+This command is sent before starting SASL authentication, and indicates that the client wishes to resume a previous connection to the server which may still be active. The command is sent with the following format:
```

Comment:
It should be before SASL authentication has completed. Once SASL auth is finished, registration completes and the client is connected normally. This command replaces the usual registration process with playing the reg info, current modes, connection to any channels the client's currently joined to and then message history if the server has it.

Sending the command after SASL auth has been completed wouldn't really work since the client would already be connected and the server would need to rename+merge the client connections in a much weirder way, which just seems sketchy imo.